src/include/OSL/Imathx/ImathMatrix.h

```diff
@@ -1133,7 +1133,7 @@ Matrix33<T>::equalWithAbsError (const Matrix33<T> &m, T e) const
 {
     for (int i = 0; i < 3; i++)
         for (int j = 0; j < 3; j++)
-            if (!IMATH_INTERNAL_NAMESPACE::equalWithAbsError ((*this)[i][j], m[i][j], e))
+            if (!IMATH_INTERNAL_NAMESPACE::equalWithAbsError ((*this).x[i][j], m.x[i][j], e))
```

Comment:
Hey, Alex, just going over this again: Is this one really necessary? Because operator[](i) is already overloaded to be x[i]. So does the insertion if this particular .x really do anything? Is the fact that the operator is a method (even if inlined?) enough to throw off the vectorization?

Reply:
```
template <class T>
IMATH_HOSTDEVICE inline T *
Matrix33<T>::operator [] (int i)
{
    return x[i];
}
```
@lgritz , It inlines, it's really the loss of type information.  Take it very literally from a compiler's point of view.  m.x[i][j] is a native access to a data member that is a 2 dimensional C array.  Compiler can track that back for SROA (Scalar Replacement of Aggregates) optimization pass.  Whereas m[i][j] is 2 operator calls, 1st returns a float * to data member, the 2nd operator subscript is an array access with an index.  As soon as a pointer is involved, compiler may not be able to track it back to a specific data member but instead do the 2 step grab pointer, then dereference pointer.  The bigger implication is the data layout may not be privatizable and the array access would need to be against standard data layout on the stack or memory vs. the possibility of keeping everything in registers (or SIMD registers when vectorizing).

Reply:
Gotcha, thanks. I'm porting these changes back to upstream Imath and wanted to be sure it all served a purpose.

Reply:
For my own Matrix classes, I don't even use arrays at all, just 9(3x3) or 16(4x4) data members and only support array subscript access with ConstIndex (std::integral_constant<int, #>) that is overloaded to access individual data members.  Basically removes ability for aliasing and pointer access that could get in the way of optimizations.

Reply:
Use of memset and memcpy in constructors might be another area of concern/tuning.  I think in my versions I directly set/copy all data members and let the compiler decide if it wants to replace it with something.  Main point is to avoid taking address of object which would then require standard data layout and an actual stack representation of something that might of been able to kept in a private data layout (SOA) in registers and only spilling registers as needed vs. the entire object.